src/core/Cx.ml

```diff
@@ -0,0 +1,249 @@
+
+(* The way that we model dimensions is now incompatible with the union-find version of things.
+   We need to find a new way. *)
+type cx =
+  {sign : GlobalEnv.t;
+   hyps : hyp list;
+   env : Domain.env;
+   qenv : Quote.env;
+   ppenv : Pretty.env;
+   rel : Restriction.t;
+   all_locks : int}
+
+type t = cx
+
+let clear_locals cx =
+  {cx with
+   qenv = Quote.Env.emp;
+   hyps = [];
+   ppenv = Pretty.Env.emp;
+   env = Domain.Env.clear_locals cx.env;
+   all_locks = 0}
+
+let hyp_map_lock f hyp =
+  {hyp with locks = f hyp.locks}
+
+
+let ext_lock cx =
+  {cx with
+   sign = GlobalEnv.ext_lock cx.sign;
+   hyps = List.map (hyp_map_lock (fun n -> n + 1)) cx.hyps;
+   all_locks = cx.all_locks + 1}
+
+let clear_locks cx =
+  {cx with
+   sign = GlobalEnv.clear_locks cx.sign;
+   hyps = List.map (hyp_map_lock (fun _ -> 0)) cx.hyps;
+   all_locks = 0}
+
+let kill_from_tick cx tgen =
+  match tgen with
+  | `Lvl (_, lvl) ->
+    let i = Quote.Env.len cx.qenv - lvl - 1 in
+    let go j hyp =
+      if j <= i then
+        {hyp with killed = true}
+      else
+        hyp
+    in
+    {cx with hyps = List.mapi go cx.hyps}
+  | `Global _ ->
+    failwith "TODO: implement kill_from_tick for global tick"
+
+let ext cx ~nm ty sys =
+  let n = Quote.Env.len cx.qenv in
+  let var = Domain.Node {con = Domain.Reflect {ty; neu = Domain.Lvl (nm, n); sys}; action = I.idn} in
+  {cx with
+   env = Domain.Env.push (Domain.Val var) cx.env;
+   hyps = {classifier = `Ty ty; locks = 0; killed = false} :: cx.hyps;
+   qenv = Quote.Env.succ cx.qenv;
+   ppenv = snd @@ Pretty.Env.bind nm cx.ppenv},
+  var
+
+let ext_tick cx ~nm =
+  let n = Quote.Env.len cx.qenv in
+  let tick = Domain.TickGen (`Lvl (nm, n)) in
+  {cx with
+   env = Domain.Env.push (Domain.Tick tick) cx.env;
+   hyps = {classifier = `Tick; locks = 0; killed = false} :: cx.hyps;
+   qenv = Quote.Env.succ cx.qenv;
+   ppenv = snd @@ Pretty.Env.bind nm cx.ppenv},
+  tick
+
+let ext_ty cx ~nm ty =
+  ext cx ~nm ty []
+
+let def cx ~nm ~ty ~el =
+  let face = Face.True (`Dim0, `Dim1, el) in
+  fst @@ ext cx ~nm ty [face]
+
+let ext_dim cx ~nm =
+  let x = Name.named nm in
+  {cx with
+   env = Domain.Env.push (Domain.Atom (`Atom x)) cx.env;
+   hyps = {classifier = `I; locks = 0; killed = false} :: cx.hyps;
+   qenv = Quote.Env.abs cx.qenv @@ Emp #< x;
+   ppenv = snd @@ Pretty.Env.bind nm cx.ppenv},
+  x
+
+let rec ext_dims cx ~nms =
+  match nms with
+  | [] -> cx, []
+  | nm::nms ->
+    (* TODO: is this backwards? *)
+    let cx, xs = ext_dims cx ~nms in
+    let cx, x = ext_dim cx ~nm in
+    cx, x :: xs
+
+let ppenv cx =
+  cx.ppenv
+
+let make_closure cx bnd =
+  Domain.Clo {rho = cx.env; bnd}
+
+let lookup i {hyps; _} =
+  let {classifier; locks; killed} = List.nth hyps i in
+  if (killed || locks > 0) && classifier != `I then
+    failwith "Hypothesis is inaccessible (modal, taste it!)"
+  else
+    classifier
+
+let lookup_constant nm tw cx =
+  if cx.all_locks > 0 then
```

Comment:
This may not be necessary; i guess somehow every constant is a global element, and might as well have been defined inside the box modality.

Reply:
Well, it's not quite as simple as that, but there is probably a story.